packages/pcd/gpc-pcd/src/validatorChecks.ts

```diff
@@ -0,0 +1,158 @@
+import { GPCProofConfig, PODMembershipLists } from "@pcd/gpc";
+import {
+  PODEntries,
+  PODName,
+  applyOrMap,
+  decodePublicKey,
+  podValueHash
+} from "@pcd/pod";
+import { PODPCD, PODPCDTypeName } from "@pcd/pod-pcd";
+import _ from "lodash";
+import {
+  PODEntryRecord,
+  PODPCDArgValidatorParams,
+  PODSignerPublicKeys
+} from "./GPCPCD";
+
+export function checkPCDType(podPCD: PODPCD): boolean {
+  return podPCD.type === PODPCDTypeName;
+}
+
+export function checkPODEntriesAgainstProofConfig(
+  podName: PODName,
+  podPCD: PODPCD,
+  proofConfig: GPCProofConfig | undefined,
+  params: PODPCDArgValidatorParams
+): proofConfig is GPCProofConfig {
+  // Check POD against serialised proof config if it is given.
+  if (proofConfig !== undefined) {
+    // POD podName should be present in the config and have all
+    // entries specified there.
+    const podConfig = proofConfig.pods[podName];
+    if (podConfig === undefined) {
+      params.notFoundMessage = `The proof configuration does not contain this POD.`;
+      return false;
+    }
+    const configEntries = Object.keys(podConfig.entries);
+
+    // Enumerate POD entries
+    const podEntries = podPCD.pod.content.asEntries();
+
+    // Return false if some entry in the config is not in the POD
+    if (
+      configEntries.some((entryName) => podEntries[entryName] === undefined)
+    ) {
+      return false;
+    } else {
+      return true;
+    }
+  } else {
+    params.notFoundMessage =
+      "The proof configuration has not been passed to the validator!";
+    return false;
+  }
+}
+
+export function checkPODEntriesAgainstMembershipLists(
+  podName: PODName,
+  podPCD: PODPCD,
+  proofConfig: GPCProofConfig,
+  membershipLists: PODMembershipLists | undefined
+): boolean {
+  // Bypass if there are no membership lists.
+  if (membershipLists === undefined) {
+    return true;
+  }
+
+  const podConfig = proofConfig.pods[podName];
+  const podEntries = podPCD.pod.content.asEntries();
+
+  // Record mapping entry names to membership list names.
+  const listMembershipRecord = Object.fromEntries(
+    Object.entries(podConfig.entries).flatMap(([entryName, entryConfig]) =>
+      entryConfig.isMemberOf !== undefined
+        ? [[entryName, entryConfig.isMemberOf]]
+        : []
+    )
+  );
+
+  // If there are list membership checks in the proof config *and* the
+  // serialised lists are passed in as a parameter, check whether any
+  // do not pass.
+  return (
+    Object.keys(listMembershipRecord).length === 0 ||
+    Object.entries(listMembershipRecord).every(([entryName, listName]) =>
+      membershipLists[listName]
+        .map((value) => applyOrMap(podValueHash, value))
```

Comment:
This is potentially calculating a lot of hashes if the list is large, and there are lots of PODs to be validated.  Thats a general problem with the structure of the vlalidation function, but lists are a point where there's an extra multiplier.

That's probably a secondary concern for now, but we may have to consider memoization at some point to avoid duplicate calculations.  Meanwhile I might suggest reordering the checkers to put the most expensive one last.  (See separate comment above.)